### PR TITLE
Probabilistic sampler processor based on draft t-value/r-value encoding

### PR DESCRIPTION
**Pull request by jmacd**
**Description:** 
Draft for https://github.com/open-telemetry/opentelemetry-specification/issues/3602.
Previously https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/22058, see also https://github.com/open-telemetry/oteps/pull/235

Reviewers: thank you for your patience. While this has was first started in May 2023 and has progressed slowly, it's finally ready and the Sampling SIG has converged on this design.

This PR adds two consistent sampler modes to the existing probabilistic sampler, "equalizing" and "proportional". The consistent sampling mechanism used here is based on 56 bits of randomness which is extracted from the tracestate or the TraceID. Because this is a tail sampler, there is no real benefit to synthesizing an r-value, therefore when `rv:` is not set in the tracestate, this implementation unconditionally uses the TraceID bits.

When https://github.com/open-telemetry/opentelemetry-proto/pull/503 is included in the next release of OTLP and pdata is extended with the corresponding field, this processor will be updated to use the randomness bit before deciding what to do about TraceIDs that do not have it set. We imagine adding a configuration parameter to indicate how to treat this scenario, but since this is a tail processor, there is little we can do that would be consistent, so using the TraceID bits as they are is probably best, despite the missing confirmation of their randomness.

Note there is a question of how to treat spans that arrive with an empty tracestate, indicating what is likely to be 100% sampling but could potentially be otherwise. There are no OpenTelemetry specifications for what we should do in this situation, and we leave this decision open.

**Link to tracking Issue:** 
https://github.com/open-telemetry/opentelemetry-specification/issues/1413

**Testing:** Complete.

**Documentation:** Added new documentation.

### CODE REVIEW

**Comment by jmacd** [comment]
Based on the feedback received in Sampling SIG today, I refactored the implementation to avoid the use of floating point numbers except when explicitly converting thresholds to probabilities and back to thresholds. I have also added a second implementation of the based on byte arrays instead of unsigned numbers, for comparison. The unsigned implementation is trickier to read and write, but faster.

```
goos: darwin
goarch: arm64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/pkg/sampling
BenchmarkThresholdCompareAsUint64
BenchmarkThresholdCompareAsUint64-10    	1000000000	         0.7034 ns/op	       0 B/op	       0 allocs/op
BenchmarkThresholdCompareAsBytes
BenchmarkThresholdCompareAsBytes-10     	191687650	         5.990 ns/op	       0 B/op	       0 allocs/op
```

**Comment by jpkrohling** [comment]
@jmacd , is this ready for a first review?

File: pkg/sampling/internal/bytes/randomness.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	"go.opentelemetry.io/collector/pdata/pcommon"
+)
+
+// ErrRValueSize is returned for r-values != NumHexDigits hex digits.
+var ErrRValueSize = errors.New("r-value must have 14 hex digits")
+
+// Randomness may be derived from r-value or TraceID.
+type Randomness struct {
+	// bytes[0] is unused, so that the relevant portion of these 8
+	// bytes align with the TraceID's second 8 bytes.
```
Comment (PeterF778) [inline pkg/sampling/internal/bytes/randomness.go:16]: What is the benefit of aligning the indices? It looks like you use array slices for any comparisons.

File: pkg/sampling/internal/bytes/threshold.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+var (
+	// ErrTValueSize is returned for t-values longer than NumHexDigits hex digits.
+	ErrTValueSize = errors.New("t-value exceeds 14 hex digits")
+
+	NeverSampleThreshold = Threshold{bytes: [8]byte{
+		0, 0, 0, 0, 0, 0, 0, 0,
+	}}
+	AlwaysSampleThreshold = Threshold{bytes: [8]byte{
+		1, 0, 0, 0, 0, 0, 0, 0,
```
Comment (PeterF778) [inline pkg/sampling/internal/bytes/threshold.go:48]: It seems the 'bytes' for NeverSampleThreshold and AlwaysSampleThreshold are never referenced. If this is true, there's no need for keeping 8 bytes. Am I missing something?

**Comment by jmacd** [comment]
This PR is nearly ready to review.

**Comment by jmacd** [comment]
Reviewer notes: I am returning this PR to draft mode. I will copy the `pkg/sampling` package out of this PR and form a new PR with it. After that stuff merges (if we may), then I'll re-open this one.

**Comment by jmacd** [comment]
Replaced by #31894.

**Pull request closed without merging**